phangsPipeline/pipelineLogger.py

```diff
@@ -0,0 +1,178 @@
+
+import logging, os
+
+DefaultLevel = 'DEBUG'
+
+try:
+    from . import casaStuff
+    HasCasaLog = True
+except:
+    HasCasaLog = False
+
+
+class PipelineLogger(logging.getLoggerClass()):
+    """A smart logging module which will redirect to CASA logger if in CASA otherwise to screen and file."""
+    def __init__(self, name, level=None, logfile=None):
+        self.name = name
+        self.origin = ''
+        self.screen_log_format = '[%(levelname).4s] [%(funcName)25s] %(message)s'
+        self.file_log_format = '[%(asctime)-15s] [%(levelname)08s]  [%(name)s] [%(funcName)s] %(message)s'
+        self.file_handler = None
+        super(PipelineLogger, self).__init__(name)
+        self.setup(name, level, logfile)
+    
+    def __enter__(self):
+        #print('PipelineLogger.__enter__')
+        return self
+    
+    def __exit__(self, type, value, traceback):
+        #print('PipelineLogger.__exit__')
+        if self.file_handler is not None:
+            self.file_handler.close()
+    
+    def __del__(self):
+        #print('PipelineLogger.__del__')
+        if self.file_handler is not None:
+            self.file_handler.close()
+    
+    def hasCasaLog(self):
+        global HasCasaLog
+        return HasCasaLog
+        #return ('casalog' in globals())
+    
+    def setCasaOrigin(self):
+        if self.hasCasaLog():
+            self.origin = 'casa'
+            casaStuff.casalog.origin(self.name)
+    
+    def restoreCasaOrigin(self):
+        if self.hasCasaLog():
+            casaStuff.casalog.origin(self.origin)
+    
+    def setup(self, name, level=None, logfile=None):
+        self.name = name
+        if not self.hasCasaLog():
+            if level is None:
+                level = DefaultLevel
+            self.setLevel(logging.getLevelName(level))
+            consoleHandler = logging.StreamHandler()
+            consoleHandler.setFormatter(logging.Formatter(self.screen_log_format))
+            self.addHandler(consoleHandler)
+            if logfile is not None:
+                fileHandler = logging.FileHandler(logfile, mode='a')
+                fileHandler.setFormatter(logging.Formatter(self.file_log_format))
+                self.addHandler(fileHandler)
+                self.file_handler = fileHandler
+    
+    def debug(self, message):
+        if self.hasCasaLog():
+            self.setCasaOrigin()
+            casaStuff.casalog.post(message, 'NORMAL')
```

Comment:
Does casa have a debug level in the logger?

Reply:
Right, it should be set to `'DEBUGGING'` (or `DEBUG1` or `DEBUG2`), according to [https://casa.nrao.edu/casadocs/casa-6.0/usingcasa/casa-logger](https://casa.nrao.edu/casadocs/casa-6.0/usingcasa/casa-logger).